[feature][PolyhedralGrid] Make PolyhedralGrid fully coherent with the DUNE grid interface.

dr-robertk opened the pull request:
This PR make the implementation of PolyhedralGrid fully compliant with DUNE grid interface and enables DUNE grid checks. Also, some minor compilation errors are fixed for CpGrid and current DUNE master.

dr-robertk commented:
jenkins build this please

atgeirr commented:
Do you want any reviewing or checking of the PolyhedralGrid parts? The CpGrid-related parts look fine.

dr-robertk commented:
If you want to spend time that would be good, but I understand it's a big PR and I just squashed things together. This PR basically added a lot of things I had implemented over the last years in various different branches. The fact that test_polyhedralgrid passed the DUNE grid check gives enough confidence and that the implementation is correct. What may be interesting is the Geometry implementation where I use the DUNE geometries to get the local-global mappings. That may be something for CpGrid also.

dr-robertk commented:
jenkins build this please

atgeirr commented:
> The fact that test_polyhedralgrid passed the DUNE grid check gives enough confidence and that the implementation is correct.

I agree that gives some confidence.

> What may be interesting is the Geometry implementation where I use the DUNE geometries to get the local-global mappings. That may be something for CpGrid also.

Yes, that is interesting. I am curious: how did you deal with the generic case when a cell has no reasonable reference element? How do you distinguish between those cases? (I'd be happy to read the code, given a pointer.)

dr-robertk commented on opm/grid/polyhedralgrid/geometry.hh:
```diff
@@ -41,6 +45,70 @@ namespace Dune
 
+    typedef typename Grid::Traits::ExtraData  ExtraData;
+    typedef typename Grid::Traits::template Codim<codimension>::EntitySeed EntitySeed;
+
```
@atgeirr : You can start reading from here.

atgeirr replied:
Thanks, I looked at it and saw some things that were new to me. I did notice a minor possible bug, will make a note in the diff.

atgeirr commented on opm/grid/polyhedralgrid/geometry.hh:
```diff
@@ -41,6 +45,70 @@ namespace Dune
+          : data_( data ), seed_()
+        {}
+
+        ExtraData data() const { return data_; }
```
Should this rather return a const reference? I imagine ExtraData is not something you want to copy for every call to corner() or center()?

dr-robertk replied:
ExtraData is of type Grid* so this is no problem (see grid.hh). EntitySeed is essentially the element index or similar so copying this is also fine.

atgeirr replied:
Ok, if you want this merged now that is fine with me.

atgeirr commented:
jenkins build this please

dr-robertk commented:
It's ok from my side and would nice if it could be merged.

akva2 commented:
this broke build on ubuntu xenial (dune 2.4), ref https://ci.opm-project.org/job/opm-nightly-xenial/164/console
